File: year2022/day04/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: number[][][] = file.toString().trim().split('\n').map((group) => group.split(',').map(tupel => tupel.split('-').map(Number)));
if(process.argv[2] === 'puzzle') {
    console.log(part_1());
    console.log(part_2());
}

export function part_1() {
    return input.filter(([pair1, pair2]) => overlaps(pair1, pair2)).length;
  }
  
export function part_2() {
    return input.filter(([pair1, pair2]) => contains(pair1, pair2)).length;
}
  
function overlaps(pair1, pair2) {
    return ((pair1[0] <= pair2[0] && pair1[1] >= pair2[1]) || (pair2[0] <= pair1[0] && pair2[1] >= pair1[1]));
}
  
function contains(pair1, pair2) {
    return ((pair1[0] >= pair2[0] && pair1[0] <= pair2[1]) || (pair2[0] >= pair1[0] && pair2[0] <= pair1[1]) || (pair1[1] <= pair2[1] && pair1[1] >= pair2[0]) || (pair2[1] <= pair1[1] && pair2[1] >= pair1[0]));
}

File: year2022/day04/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return len(list(filter(lambda pair: overlaps(pair[0], pair[1]), input)))


def part_2():
  return len(list(filter(lambda pair: contains(pair[0], pair[1]), input)))


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [list(map(lambda tupel: list(map(int, tupel.split('-'))), group.split(','))) for group in file.read().split('\n')]


def overlaps(pair1, pair2):
  return (pair1[0] <= pair2[0] and pair1[1] >= pair2[1]) or (pair2[0] <= pair1[0] and pair2[1] >= pair1[1])


def contains(pair1, pair2):
  return ((pair1[0] >= pair2[0] and pair1[0] <= pair2[1]) or (pair2[0] >= pair1[0] and pair2[0] <= pair1[1]) or (pair1[1] <= pair2[1] and pair1[1] >= pair2[0]) or (pair2[1] <= pair1[1] and pair2[1] >= pair1[0]))


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()